Create hands on introduction to git & github workshop material. (#20)
Arcadia-Science/arcadia-computational-training: docs/workshops/20220920-intro-to-git-and-github/lesson.md

You can always go back to a previous commit.

**GitHub** is an internet service for hosting files that are under git version control.
You can think of it like social media for code and small text files.
GitHub houses repositories in a central locations.
Repositories on GitHub are referred to as **remote** repositories.
Users can create a **local** copy of a repository by **cloning** (downloading) it to their computer, or to a computer in the cloud that they have access to.

You don't need GitHub to use the Git version control system.
By using the Git command line interface, you can use version control locally without ever **pushing** your files to a remote repository on GitHub. 

You also don't need to know the Git command line interface to take advantage of the version control system: Git is well-integrated into GitHub, so if you know the basic terminology and definitions, you can still version control your project through GitHub alone.

However, both Git and GitHub are most powerful when combined together.
This workshop will teach you the basics of Git and GitHub so you can leverage both in your research.

## Setting up

> #### Signing up for a GitHub account

Author: it looked nicer when it rendered. going to `###` made the subheaders too big and it was harder to follow the doc within a subsection when it was rendered. I can change it though if you think that's better for consistency

If Git is not already installed on your computer, see [these instructions](https://carpentries.github.io/workshop-template/#git).

#### First time setup and configuration

Whenever you use Git for the first time on a computer, there are a few things you need to set up.
Unless you need to change something, these commands only need to be used the first time you use Git on a computer.

First, set your name and email:
```
git config --global user.name "Taylor Reiter"
git config --global user.email "myemail@gmail.com"
```

> Your email and user name is recorded with every commit.
> This helps ensure integrity and authenticity of the history.
> Most people keep their email public, but if you are concerned about privacy, check GitHub's tips to [hide your email](https://help.github.com/articles/keeping-your-email-address-private/).

Next, we'll set up a key pair. 
GitHub [recently discontinued password authentication](https://docs.github.com/en/rest/overview/other-authentication-methods). 
Cryptographic keys are a convenient and secure way to authenticate without having to use passwords and are an authentication method still supported by GitHub. 

> They consist of a pair of files called the public and private keys: the public part can be shared with whoever you'd like to authenticate with (in our case, GitHub), and the private part is kept "secret" on your machine. 

Author: Ooh good call
```suggestion
They consist of a pair of files called the public and private keys: the public part can be shared with whoever you'd like to authenticate with (in our case, GitHub), and the private part is kept "secret" on your machine. 
The private key should never be shared in an unencrypted way with anyone -- this includes over email, slack, etc. 
If your private key is accidentally shared, you should stop using it for authentication and create a new key pair.
```

```

To upload the key file text, navigate to GitHub and click on settings.

<center>
![](github_settings.png){ width="300" }
</center>

Then use the menu on the left hand side of the page to navigate to the `SSH and GPG keys` tab.

<center>
![](github_ssh_tab.png){ width="300" }
</center>

Once there, select `New SSH key`.

<center>
![](github_new_ssh.png){ width="500"}
</center>

Give your key a descriptive name (such as `20220909-github-workshop`) and then paste in the contents of your public key file to text editor.

<center>
![](github_ssh.png){ width="900" }
</center>

The very last thing we need to do is tell our computers which key file to use when we want to authenticate with GitHub.
We do this by creating a `config` file in our `.ssh` directory.
We'll use nano to do this.
Type the following contents into your `config` file and save it as `config`.

```
nano config
```

```
Host github.com
 User git
 HostName ssh.github.com
 IdentityFile ~/.ssh/20220920-github-workshop
```

> Follow the command prompts in the bottom ribbon of nano to exit.

Author: ooh good call. will add.

1. Navigate to [GitHub](www.github.com).
2. Click on the plus sign in the upper right of the screen.
3. Select "New repository" from the drop-down menu.
4. On the "Create a new repository" page, choose a name for your repository.
   For this workshop, name the repository `2022-git-workshop`. 
   A repository can have any valid directory name, but putting the year at the beginning is a good practice because it makes it clear when the repo was created.
5. Check the box "Add a README file".
6. Click the green "Create repository" button at the bottom of the page.

This will create a new repository and take you to the repository's landing page on GitHub.

#### Cloning a repository from remote to local

**Cloning** is the process of copying an existing Git repository from a **remote** location (here, on GitHub) to your **local** computer (here, on our binder instances).
To clone the repository, click the green "Code" button in the top right hand corner of the repository screen. 
This creates a drop down menu with clone options.
We'll select the SSH tab because we configured an ssh key pair.
Once you select the tab, copy the path that starts with `git@gitub.com:`.

> Then, navigate to your binder instance and use the command below to clone the repository.

Author: yes good catch thank you! 
```suggestion
Then, navigate to your terminal and use the command below to clone the repository.
```

Remember to substitute out the username/URL with your own URL that we copied.

```
cd ~
git clone git@github.com:your_username/2022-git-workshop
cd 2022-git-workshop
```

#### The basic Git workflow

<center>
<figure markdown>
  ![Image title](allison_horst_git_overview_small.png){ width="600" }
  <figcaption> Figure by Allison Horst. <a href='https://twitter.com/allison_horst/status/1523744022421860352' target='_blank'>www.twitter.com/allison_horst</a> </figcaption>
</figure>
</center>

We can now make changes to our files in our local repository.
The basic Git workflow begins when you communicate to Git about changes you've made to your files.
Once you like a set of changes you’ve made you can tell Git about them by using the command **`git add`**.

> This stages the changes you have made.

Author: ```suggestion
This stages the changes you have made.
Staging a file tells Git that you're ready to commit those files -- it's a way of telling Git which files you're ready to commit.
```

Next, you bake them into the branch using **`git commit`**.
When you’re ready, you can communicate those changes back to GitHub using **`git push`**.
This will push your the changes that are in your local repository up to the remote repository.

Let's try this workflow out.
Throughout this process, we'll use the command `git status` to track our progress through the workflow.
`git status` displays the state of a working direcotry and the staging area.

```
git status
```

We'll use the `echo` command to create a new file, `notes.txt`.

```
ls
echo "some interesting notes" > notes.txt
ls
```

Take a look at the contents of your `notes.txt` file:
```
less notes.txt
```

And run `git status` to see how creating a new file changes the output of that command:
```
git status
```

Once you have made changes in your repository, you need to tell Git to start tracking the file. 
The command `git add` adds the files to the "staging area", meaning Git is now tracking the changes.

```

> git add notes.txt

Author: Ya i see how that could be useful after running `git status`, but I'd rather not teach it at first -- I think it could lead to people accidentally forgetting the sequence to run things, and then lead to files accidentally getting added to repos (e.g. like `.DS_store`). Maybe a month or so after this workshop we could do an AUG on efficiency with git and github if we see an uptick in the number of people committing to Arcadia-Science org?

```

After adding this file, we see our output of `git status` changes.

```
git status
```

The text associated with our file is now green because the file is staged.
When you've made all of the changes to a file that represent a unit of changes, you can use `git commit` to create a snapshot of the file in your Git version history.

```
git commit -m "start notes file"

git status
```

The `-m` flag is the message that is associated with that commit.
The message should be short and descriptive so that you or someone looking at your code could quickly determine what changes took place from one commit to the next.

> What constitutes a unit of changes worthy of running `git commit`? 

Author: no, I was keeping it as part of the basic git workflow heading. Do you think it should be a separate markdown header?

> What constitutes a unit of changes worthy of running `git commit`? 

Reviewer: It's fine

In our branch, we'll update our `README.md` file.
Use the code below to create a new branch, but remember to substitute out `your-initials` for your initials.

```
git checkout -b your-initials/update-readme.
```

Whenever you create a new branch, it branches off from the branch you're currently in when you make the new branch.
In this case, we started in `main`, so our branch will branch off of from here.

Next, let's add some changes to our `README.md` file and run `git status`.

```
echo "adding more text to the README file" >> README.md
git status
```

Stage the changes with `git add`:

```
git add README.md
```

And commit them:
```
git commit -m "update readme w more text"
```

Run `git push`:

```
git push
```

This gives us an error message!

```
fatal: The current branch tmp has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin ter/update-readme
```

While we created our branch locally, we haven't created the same branch remotely. 
We need to tell `git` where to put our changes in the remote repository.
Luckily, the error message points us to the code we need to run to set the remote branch.

```

> git push --set-upstream origin ter/update-readme

Author: ooh tell me more! I know nothing about this setting! I think that is desirable -- not sure if it would best be delivered in this workshop vs. in a later AUG efficiency with git. We could always add it to this lesson material and I could skip covering it if we're short on time

> git push --set-upstream origin ter/update-readme

Reviewer: I think it's a config:
```
git config --global push.default current
git config --global remote.pushDefault origin
```
